File: app/src/main/java/com/worldventures/dreamtrips/modules/tripsimages/service/command/SendAnalyticsIfNeedAction.java
package com.worldventures.dreamtrips.modules.tripsimages.service.command;

import com.worldventures.dreamtrips.core.janet.dagger.InjectableAction;
import com.worldventures.dreamtrips.core.utils.tracksystem.AnalyticsInteractor;
import com.worldventures.dreamtrips.modules.infopages.service.analytics.WatchVideoAnalyticAction;
import com.worldventures.dreamtrips.modules.reptools.presenter.TrainingVideosPresenter;
import com.worldventures.dreamtrips.modules.video.presenter.HelpVideosPresenter;
import com.worldventures.dreamtrips.modules.video.presenter.PresentationVideosPresenter;

import javax.inject.Inject;

import io.techery.janet.Command;
import io.techery.janet.command.annotations.CommandAction;

@CommandAction
public class SendAnalyticsIfNeedAction extends Command<Integer> implements InjectableAction {

   @Inject AnalyticsInteractor analyticsInteractor;

   private final Class launchComponent;
   private final String language;
   private final String videoName;

   private int expectedAnalyticStep;
   private final long currentVideoProgress;
   private final long totalVideoLength;

   public SendAnalyticsIfNeedAction(Class launchComponent, String language, String videoName, int expectedAnalyticStep, long currentVideoProgress, long totalVideoLength) {
      this.launchComponent = launchComponent;
      this.language = language;
      this.videoName = videoName;
      this.expectedAnalyticStep = expectedAnalyticStep;
      this.currentVideoProgress = currentVideoProgress;
      this.totalVideoLength = totalVideoLength;
   }

   @Override
   protected void run(CommandCallback<Integer> callback) throws Throwable {
      int percent = (int) ((currentVideoProgress * 100) / totalVideoLength);
      int step = percent / 25;
      WatchVideoAnalyticAction action = chooseAnalyticAction(step, expectedAnalyticStep);
      if (action != null) {
         analyticsInteractor.analyticsActionPipe().send(action);
         callback.onSuccess(expectedAnalyticStep + 1);
      } else
         callback.onSuccess(expectedAnalyticStep);
   }

   private WatchVideoAnalyticAction chooseAnalyticAction(int analyticStep, int currentAnalyticStep) {
      if (currentAnalyticStep == 0)
         return WatchVideoAnalyticAction.startVideo(language, videoName, chooseAnalyticNamespace());

      if (analyticStep < currentAnalyticStep) return null;

      WatchVideoAnalyticAction action = null;
      switch (analyticStep) {
         case 1:
            action = WatchVideoAnalyticAction.progress25(language, videoName, chooseAnalyticNamespace());
            break;
         case 2:
            action = WatchVideoAnalyticAction.progress50(language, videoName, chooseAnalyticNamespace());
            break;
         case 3:
            action = WatchVideoAnalyticAction.progress75(language, videoName, chooseAnalyticNamespace());
            break;
         case 4:
            action = WatchVideoAnalyticAction.progress100(language, videoName, chooseAnalyticNamespace());
      }
      return action;
   }

   private String chooseAnalyticNamespace() {
      if (launchComponent.equals(HelpVideosPresenter.class)) return WatchVideoAnalyticAction.HELP_VIDEO_NAMESPASE;
      if (launchComponent.equals(PresentationVideosPresenter.class))
         return WatchVideoAnalyticAction.MEMBERSHIP_VIDEOS_NAMESPASE;
      if (launchComponent.equals(TrainingVideosPresenter.class))
         return WatchVideoAnalyticAction.REPTOOLS_TRAINING_VIDEOS_NAMESPASE;
      return null;
   }

}

File: app/src/main/java/com/worldventures/dreamtrips/social/ui/tripsimages/service/analytics/SendAnalyticsIfNeedAction.kt
package com.worldventures.dreamtrips.social.ui.tripsimages.service.analytics

import com.worldventures.core.service.analytics.AnalyticsInteractor
import com.worldventures.dreamtrips.social.ui.infopages.service.analytics.WatchVideoAnalyticAction
import com.worldventures.dreamtrips.social.ui.reptools.presenter.TrainingVideosPresenter
import com.worldventures.dreamtrips.social.ui.video.presenter.HelpVideosPresenter
import com.worldventures.dreamtrips.social.ui.video.presenter.PresentationVideosPresenter
import com.worldventures.janet.injection.InjectableAction
import io.techery.janet.Command
import io.techery.janet.command.annotations.CommandAction
import javax.inject.Inject

@CommandAction
class SendAnalyticsIfNeedAction(private val launchComponent: Class<*>, private val language: String,
                                private val videoName: String, private val expectedAnalyticStep: Int,
                                private val currentVideoProgress: Long, private val totalVideoLength: Long) : Command<Int>(), InjectableAction {

   @Inject internal lateinit var analyticsInteractor: AnalyticsInteractor

   @Throws(Throwable::class)
   override fun run(callback: Command.CommandCallback<Int>) {
      val percent = (currentVideoProgress * 100 / totalVideoLength).toInt()
      val step = percent / PERCENT_STEP
      val action = chooseAnalyticAction(step, expectedAnalyticStep)
      if (action != null) {
         analyticsInteractor.analyticsActionPipe().send(action)
         callback.onSuccess(expectedAnalyticStep + 1)
      } else {
         callback.onSuccess(expectedAnalyticStep)
      }
   }

   private fun chooseAnalyticAction(analyticStep: Int, currentAnalyticStep: Int): WatchVideoAnalyticAction? {
      if (currentAnalyticStep == 0) {
         return WatchVideoAnalyticAction.startVideo(language, videoName, chooseAnalyticsNamespace())
      }

      if (analyticStep < currentAnalyticStep) {
         return null
      }

      var action: WatchVideoAnalyticAction? = null
      when (analyticStep) {
         PROGRESS_25_STEP -> action = WatchVideoAnalyticAction.progress25(language, videoName, chooseAnalyticsNamespace())
         PROGRESS_50_STEP -> action = WatchVideoAnalyticAction.progress50(language, videoName, chooseAnalyticsNamespace())
         PROGRESS_75_STEP -> action = WatchVideoAnalyticAction.progress75(language, videoName, chooseAnalyticsNamespace())
         PROGRESS_100_STEP -> action = WatchVideoAnalyticAction.progress100(language, videoName, chooseAnalyticsNamespace())
         else -> {
         }
      }
      return action
   }

   private fun chooseAnalyticsNamespace(): String? {
      return when (launchComponent) {
         HelpVideosPresenter::class.java -> WatchVideoAnalyticAction.HELP_VIDEO_NAMESPASE
         PresentationVideosPresenter::class.java -> WatchVideoAnalyticAction.MEMBERSHIP_VIDEOS_NAMESPASE
         TrainingVideosPresenter::class.java -> WatchVideoAnalyticAction.REPTOOLS_TRAINING_VIDEOS_NAMESPASE
         else -> null
      }
   }

   companion object {
      const val PERCENT_STEP = 25

      const val PROGRESS_25_STEP = 1
      const val PROGRESS_50_STEP = 2
      const val PROGRESS_75_STEP = 3
      const val PROGRESS_100_STEP = 4
   }
}
